Upgrade to Publishing Api V2

davidbasalla opened the pull request:
This PR switches the app to using Publishing API V2. This includes a new `Publisher` class which bundles the three calls (`put_content`, `put_links` and `publish`) that need to be made with the new API to publish an item to 'live'. All tests have been updated to use this new `Publisher` class as well as to use the V2 relevant test helpers.

Contract testing has also been upgraded to V2, which required an update to the ContentItem Presenters, since `content_id` and `update_type` are now no longer part of the payload, but need to passed as separate parameters.

Finally, any code regarding the finder has been removed, as the finder has never been implemented fully. 

For https://trello.com/c/An50dl6N/213-contacts-bump-from-v1-to-v2-publishing-api

issyl0 commented:
Thank you for working on this in my 2ndline-induced absence! :tada: I'll take a look in detail soon.

issyl0 commented on app/presenters/contacts_finder_presenter.rb:
```diff
@@ -1,105 +0,0 @@
-class ContactsFinderPresenter
```
Will removing this file negatively affect the contacts index page? I guess not, 'cause you've removed it and checked it all still works, but just for peace of mind.

davidbasalla replied:
I checked the HMRC contact index page on dev and everything is still working as expected. This ContactsFinderPresenter was only ever called inside the [`publish_finder`](https://github.com/alphagov/contacts-admin/blob/master/app/models/contact_group.rb#L38-L41) method in contact_group.rb, which in turn was commented out.

issyl0 replied:
:+1:

issyl0 commented:
David talked me through this and it all seems sensible, and I've seen it all working as normal.

boffbowsh commented on lib/publisher.rb:
```diff
@@ -0,0 +1,44 @@
+
+  def publish
+    response = send_item_to_draft_stack
+    response = publish_links if presenter.links && response_successful?(response)
```
It seems like the client methods will always raise an error in the event of a non-2XX response, so I don't think the `response_successful?` checks are needed. Also, the API might change to a more semantically correct 201 or 202 one day, which would fail under your `response_successful?` check.

boffbowsh commented on spec/lib/publisher_spec.rb:
```diff
@@ -0,0 +1,91 @@
+
+      it 'calls put_content' do
+        expect(Publisher.client).to receive(:put_content)
+        .with(contact.content_id, presenter.payload)
```
According to the styleguide as enforced by [govuk-lint](https://github.com/alphagov/govuk-lint), these should be indented. It's worth adding govuk-lint to this repo, all bar two of the 274 contraventions can be automatically corrected. See the README for implementation.

davidbasalla replied:
great, will do!

boffbowsh commented on lib/publisher.rb:
```diff
@@ -0,0 +1,44 @@
+    response = send_item_to_draft_stack
+    response = publish_links if presenter.links && response_successful?(response)
+    response = publish_item if response_successful?(response)
+    response
```
It feels like the caller of this shouldn't need to know about the response, and understand that a 200 is good and other things are bad. Looking through the rest of the PR, this return value isn't used for anything, what error handling needs to happen here?

davidbasalla replied:
I added the response as a return value because the code is checked in [redirector_for_gone_contact.rb](https://github.com/alphagov/contacts-admin/blob/upgrade-to-publishing-api-v2/lib/redirector_for_gone_contact.rb#L87-L89). That is the only place it is used though...

boffbowsh replied:
Might be better to just return `true` or `false` here then, so the caller doesn't need to know about the internals

boffbowsh commented on spec/features/admin/contact_edit_spec.rb:
```diff
@@ -36,7 +36,7 @@
                   )
 
-    assert_publishing_api_put_item(contact.link, title: "new title", description: "new description")
+    assert_publishing_api_put_content(contact.content_id, { title: "new title", description: "new description" })
```
The `{ }` aren't required here

boffbowsh commented on spec/presenters/contact_presenter_spec.rb:
```diff
@@ -4,50 +4,63 @@
+    it "returns links data" do
+      links = ContactPresenter.new(contact).links
+
+      expect(links[:links]["organisations"].length).to eq(1)
```
Should also validate these:
```ruby
expect(links).to be_valid_against_links_schema('contact')
```

boffbowsh commented on app/presenters/contact_presenter.rb:
```diff
@@ -7,35 +7,45 @@ def initialize(contact)
       locale: "en",
       publishing_app: "contacts",
       rendering_app: "contacts-frontend",
-      update_type: "major",
```
Removal of this might be invalid, see here: https://github.com/alphagov/whitehall/pull/2431#issuecomment-172898454

davidbasalla replied:
Hmm so when I revert this change, I get a failing contract [test](https://github.com/alphagov/contacts-admin/blob/fd3a89c24d07e5baf2ab201353b3038cc6d365fa/spec/presenters/contact_presenter_spec.rb#L16-L19) against the V2 contacts schema. Does the above comment suggest that the V2 schemas will change in future so as to allow `update_type` as part of the payload again?

davidbasalla replied:
Spoke to @tommyp and he will update `govuk_content_schemas` to allow for `update_type` being in the payload again

tommyp replied:
Done. https://github.com/alphagov/govuk-content-schemas/pull/208

danielroseman commented:
The linter seems to be going against some of our style guide recommendations - in particular, the guide says to leave a trailing comma at the end of a multiline hash or array, to reduce diff noise when subsequently adding extra items.

benlovell commented:
@danielroseman This rule needs to be enabled: https://github.com/alphagov/govuk-lint/blob/master/configs/rubocop/gds-ruby-styleguide.yml#L290

davidbasalla commented:
@boffbowsh @danielroseman @benlovell thanks, will address the linting issues in a different PR to make this one easier to read

davidbasalla commented:
@boffbowsh Have addressed all your comments and rebased
- `Publisher.publish` method now returns a boolean [here](https://github.com/alphagov/contacts-admin/blob/upgrade-to-publishing-api-v2/lib/publisher.rb#L12-L19)
- `update_type` now remains part of the presenter payload [as update_type is allowed in payload again](https://github.com/alphagov/govuk-content-schemas/pull/208)
- contract testing for links added
- `govuk-lint` will be run in a follow up PR

boffbowsh commented:
:+1: but will let @issyl0 check it over again as she has more domain context than I

issyl0 commented:
Thanks for the sylistic comments @boffbowsh - I've not got the hang of spotting those yet.

issyl0 commented:
If it works (which it does), I'm happy with it. :-)
